File: 39. CombinationSum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(candidates, target) {
    candidates = candidates.sort((a, b) => {
        return a - b;
    })

    const resArr = [];
    const combinationSumRe = (candidatesNum, target, start, res) => {
        for(let i = start, len = candidatesNum.length; i < len; i ++) {
            if(target == 0) {
                resArr.push(Object.assign([], res));
                return;
            }
            if(candidatesNum[i] > target) {
                break;
            }

            res.push(candidatesNum[i]);
            combinationSumRe(candidatesNum, target - candidatesNum[i], i, res);
            res.pop();
        }
    };
    combinationSumRe(candidates, target, 0, []);
    return resArr;
};

combinationSum([2,3,6,7], 7);
combinationSum([2,3,5], 8);
combinationSum([1], 1);
combinationSum([10,11,12], 8);

File: 39. Combination Sum.py
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        def combinationSumDFS(nums, target, start, out, res):
            if target < 0:
                return
            if target == 0:
                res.append(out)
                return
            for i in range(start, len(candidates)):
                combinationSumDFS(candidates, target - candidates[i], i, out + [candidates[i]], res)

        candidates.sort()
        res = []
        combinationSumDFS(candidates, target, 0, [], res)
        return res
        
            



            

            


test = Solution()

print test.combinationSum([1, 2,3,3,6,7], 5)